### PR TITLE
fs.download_dir: callback support

### PR DESCRIPTION
**Pull request by skshetry**

### CODE REVIEW

**Review by efiop — APPROVED** [review]

File: tests/func/test_fs.py [hunk 1]
```diff
@@ -476,3 +476,35 @@ def test_download_callback(tmp_dir, dvc, cloud):
...
+    [
+        pytest.lazy_fixture("azure"),
+        pytest.lazy_fixture("gs"),
+        pytest.param(
+            pytest.lazy_fixture("gdrive"),
+            marks=pytest.mark.xfail(
+                reason="https://github.com/iterative/PyDrive2/issues/136"
+            ),
+        ),
+        pytest.lazy_fixture("gdrive"),
```
Comment (skshetry) [inline tests/func/test_fs.py:493]:
```suggestion
```

File: tests/func/test_fs.py [hunk 2]
```diff
@@ -476,3 +476,35 @@ def test_download_callback(tmp_dir, dvc, cloud):
...
+    "cloud",
+    [
+        pytest.lazy_fixture("azure"),
+        pytest.lazy_fixture("gs"),
+        pytest.param(
+            pytest.lazy_fixture("gdrive"),
+            marks=pytest.mark.xfail(
+                reason="https://github.com/iterative/PyDrive2/issues/136"
+            ),
+        ),
```
Comment (skshetry) [inline tests/func/test_fs.py:492]:
```suggestion
```

File: tests/func/test_fs.py [hunk 3]
```diff
@@ -476,3 +476,30 @@ def test_download_callback(tmp_dir, dvc, cloud):
...
     assert (tmp_dir / "foo").read_text() == "foo"
+
+
+@pytest.mark.needs_internet
+@pytest.mark.parametrize(
+    "cloud",
+    [
+        pytest.lazy_fixture("azure"),
+        pytest.lazy_fixture("gs"),
+        pytest.param(
```
Comment (skshetry) [inline tests/func/test_fs.py:487]:
```suggestion
```

**Review by pared — APPROVED** [review]

File: tests/func/test_fs.py [hunk 4]
```diff
@@ -476,3 +476,34 @@ def test_download_callback(tmp_dir, dvc, cloud):
...
+        ),
+    ],
+)
+def test_download_dir_callback(tmp_dir, dvc, cloud):
+    cls, config, _ = get_cloud_fs(dvc, **cloud.config)
+    fs = cls(**config)
+    cloud.gen({"dir": {"foo": "foo", "bar": "bar"}})
+
+    callback = fsspec.Callback()
+    fs.download(cloud / "dir", tmp_dir / "dir", callback=callback)
```
Comment (skshetry) [inline tests/func/test_fs.py:505]: There is something wrong with `GDriveFileSystem` which is failing on this test (unrelated to the callbacks).

Comment (skshetry) [reply tests/func/test_fs.py:505]: Will be fixed by #6618.

**Pull request merged**